### DISCUSSION TITLE
Add INI sections to config parser.

### DISCUSSION THREAD

**Discussion by Zelaf** [bashly-framework/bashly#405] [Ideas]
I was wondering if it's possible to add sections to the INI parser so you can have a clearer config structure.

I'm currently working on a script that has a few config options available for the user so I want to be able to have sections for the config file and if they want to update file output file locations and such they can do that, but instead of having long strings like
```
command_config_option = setting
othercommand_output_location = folder
```
It would be nice to be able to do
```
[command]
config_option = setting

[othercommand]
output_location = folder
```
and if you were to call to the config file like `$(config_get command_config_option)` it would return `setting`.

If you know any better approach to this it would also be welcomed. At first I tried to use the yaml addon but realised that it can't really set the options so if I were to want a reset option or more expandability it wouldn't really work in the bigger picture.

**Comment by DannyBen** [comment]
I agree it is a good idea. It will most likely make the [library code](https://github.com/DannyBen/bashly/blob/master/lib/bashly/libraries/config/config.sh) much more complex, especially for setting values.

But I will take a closer look at it, see if I can implement.

**Comment by Zelaf** [comment]
I was able to make a very bad mockup for the config_get (With a little bit of help of AI) but I didn't save it since it barely worked. I was trying to see what the possibilities were for the config_set but I couldn't make heads or tails of it. 

I know the yaml library uses awk and sed, maybe it could potentially be easier with that since sed can modify the files.

**Comment by DannyBen** [comment]
I will try in the next few days to make something. 
I would like to make use dot syntax `config_get section.key`

**Comment by Zelaf** [comment]
That would be super nice.

```
[generate]
folder = /here
name = output_

[options]
colours = true
global = false
```
And just get the specific one with `config_get options.colours`. I'm excited for the possibilities of having more clean and user friendly config files!

**Comment by DannyBen** [comment]
This seems to be a working `config_get` with section support:

```bash
## Get a value from the config.
## Usage: result=$(config_get KEY) -or- result=$(config_get SECTION.KEY)
config_get() {
  local key=$1

  # Determine if we received SECTION.KEY or just KEY
  if [[ $key == *.* ]]; then
    IFS='.' read -r section key <<< "$key"
  else
    section=""
  fi

  local regex="^$key *= *(.+)$"
  local section_regex="^\[$section\]"
  local other_section_regex="^\["
  local value=""

  config_init

  while IFS= read -r line || [ -n "$line" ]; do
    # If SECTION was given, find it in the INI first
    if [[ $section && $found_section != true ]]; then
      [[ $line =~ $section_regex ]] && found_section=true
      continue
    fi

    # If SECTION was given and found, any other [SECTION] means we did not
    # find the requested KEY in the requested SECTION - abort.
    [[ $section && $line =~ $other_section_regex ]] && break

    # Found the value
    if [[ $line =~ $regex ]]; then
      value="${BASH_REMATCH[1]}"
      break
    fi
  done <"$CONFIG_FILE"

  echo "$value"
}
```

The `config_set` will take me more time.

**Comment by Zelaf** [comment]
A bit off topic, but do you know what the best way to approach to check if a config file exists? I want to make sure one exists with some default values before running any commands.

**Reply by DannyBen** [reply]
> do you know what the best way to approach to check if a config file exists?

I would do this:
1. `bashly add hooks` - and delete `src/before.sh` and `src/after.sh` - leaving just `initialize.sh`
2. In this file, do something like:

```bash
CONFIG_FILE=path_to_config.ini   
[[ -f $CONFIG_FILE ]] || create_default_config
````

and then of course, add `src/lib/create_default_config.sh` function that writes the content.

Another option, is to avoid checking/creating anything - just when you use `config_get`, assign a default value if it is empty.
Then, users can run without any config file, and still have defaults.

**Reply by Zelaf** [reply]
Ooh, default values is smart, that would even make so if someone would to break the config file it'll still work! Should I add the default values in the initialize script too?

**Reply by Zelaf** [reply]
Alright so currently I'm doing this i my init script:
```
generate_file_name=uptime_kuma_service_push_
generate_folder_location=push-scripts

## Load config file
CONFIG_FILE=config

## Checks for config file
[[ -f $CONFIG_FILE ]] || echo "$generate_folder_location and $generate_file_name" && exit 1
```
The echo and exit is just for testing for now, but is there a way to load all the collected variables like you can with the $eval in the yaml addon?

**Reply by DannyBen** [reply]
The below `config_get` accepts a second argument which is the default value:

```bash
## Get a value from the config.
## Usage: result=$(config_get KEY [DEFAULT_VALUE])
##        result=$(config_get SECTION.KEY [DEFAULT_VALUE])
config_get() {
  local key="$1"
  local value="$2"

  # Determine if we received SECTION.KEY or just KEY
  if [[ $key == *.* ]]; then
    IFS='.' read -r section key <<< "$key"
  else
    section=""
  fi

  local regex="^$key *= *(.+)$"
  local section_regex="^\[$section\]"
  local other_section_regex="^\["

  config_init

  while IFS= read -r line || [ -n "$line" ]; do
    # If SECTION was given, find it in the INI first
    if [[ $section && $found_section != true ]]; then
      [[ $line =~ $section_regex ]] && found_section=true
      continue
    fi

    # If SECTION was given and found, any other [SECTION] means we did not
    # find the requested KEY in the requested SECTION - abort.
    [[ $section && $line =~ $other_section_regex ]] && break

    # Found the value
    if [[ $line =~ $regex ]]; then
      value="${BASH_REMATCH[1]}"
      break
    fi
  done <"$CONFIG_FILE"

  echo "$value"
}
```

as for changing the approach so that it is all one big associative array - it is an option. It could be a good option, but will require a complete rewrite of this code.

**Comment by DannyBen** [comment]
How about this prototype?

```bash
#!/usr/bin/env bash

config_load() {
  declare -gA config

  local section=""
  local key=""
  local value=""
  local section_regex="^\[(.+)\]"
  local key_regex="^([^ =]+) *= *(.*) *$"
  local comment_regex="^;"
  
  while IFS= read -r line; do
    if [[ $line =~ $comment_regex ]]; then
      continue
    elif [[ $line =~ $section_regex ]]; then
      section="${BASH_REMATCH[1]}."
    elif [[ $line =~ $key_regex ]]; then
      key="${BASH_REMATCH[1]}"
      value="${BASH_REMATCH[2]}"
      config["${section}${key}"]="$value"
    fi
  done < config.ini
}

config_show() {
  for key in "${!config[@]}"; do 
    echo "$key=${config[$key]}"
  done
}

config_save() {
  local filename="out.ini"
  local current_section=""

  rm -f "$filename"

  for key in "${!config[@]}"; do
    IFS="." read -r section_name key_name <<< "$key"
    value="${config[$key]}"
    
    if [ "$current_section" != "$section_name" ]; then
      [[ $current_section ]] && echo >> "$filename"
      echo "[$section_name]" >> "$filename"
      current_section="$section_name"
    fi
    
    echo "$key_name = $value" >> "$filename"
  done
}

# read file to the config var
config_load

# use the values
echo "${config[section1.key]}"
echo "${config[section1.key2]}"
echo "${config[section2.key]}"
echo "${config[section2.key2]}"
echo "${config[section1.under_key]}"

# show the entire loaded config (associative array)
config_show

# update a value
config[section1.under_key]="updated"

# delete a value
unset config[section1.key2]
unset config[section2.key2]

# save the entire array back to the file
config_save

```

**Reply by Zelaf** [reply]
This one is working pretty good, however it added two sections in the out.ini

New commands
```
# read file to the config var
config_load

# use the values
echo "${config[section1.key]}"
echo "${config[section1.key2]}"
echo "${config[section2.key]}"
echo "${config[section2.key2]}"
echo "${config[section1.under_key]}"

# show the entire loaded config (associative array)
config_show

# update a value
config[section1.under_key]="updated"
config[generate.new-value]="good test lol"
config[generate.new_folder]="this/is/a/new/folder"

# delete a value
unset config[section1.key2]
unset config[section2.key2]

# save the entire array back to the file
config_save
```
config.ini
```
; config file for config-test.sh
[generate]
help = true
test = testing
```
out.ini
```
[generate]
new-value = good test lol
new_folder = this/is/a/new/folder

[section1]
under_key = updated

[generate]
help = true
```

**Reply by Zelaf** [reply]
How about the following call parameters?

`${config[section]}` will output the whole section
`${config[section.key]` will output the value of the key
`${config[section.key]="new value"` will update the value of the key

Or perhaps do something like this for the sections instead since it'll probably be a edgecase for people to want to load in the whole section at once.
`${config_show[section]`

A default value would be nice to implement as well in-case the config is miss-configured or not set.

The test config script also errors out if the config file doesn't exist.
`./config-test.sh: line 3: config.ini: No such file or directory`

But there seems to be some good progress here!

**Reply by DannyBen** [reply]
>  however it added two sections in the out.ini

I have seen this bug - I thought I fixed it. I will recheck.

> ${config[section]} will output the whole section

No. Arrays and associative arrays in bash are not first class citizens in the first place. There is no "array of arrays" or "hash of arrays".
I don't see a worthwhile path for implementing a "get all section". Your other two cases (get value and update value) already work, since this is a standard bash array.

> A default value would be nice to implement as well in-case the config

These are bash arrays - a default value can be obtained with `${config[section.key]:-"default value"}`.
I don't think there is a need to invent a new construct for this. Do you?

> The test config script also errors out if the config file

By design. The reference model is loading a file, or a YAML file, or any other type of file with any other programming language. It is the developer's responsibility to load a file that exists.

Contrary to the previous implementation - where the filename was implicit (via config var, or `config.ini`) - the filename here is explicit.

If you think otherwise, I am listening.

**Reply by Zelaf** [reply]
> These are bash arrays - a default value can be obtained with ${config[section.key]:-"default value"}.

> By design. The reference model is loading a file, or a YAML file, or any other type of file with any other programming language. It is the developer's responsibility to load a file that exists.

> Contrary to the previous implementation - where the filename was implicit (via config var, or config.ini) - the filename here is explicit.
I don't think there is a need to invent a new construct for this. Do you?

I'm sadly not versed in bash enough to understand it all sadly, but from what I gather, if you can set a default value using `${config[section.key]:-"default value"}` then that's great. If the filename is now explicit I think that'll work out too for better implementations since if you're using configs anyways, the options should either be put in place already through a generated config file with default values.

Currently I think I will implement a check for a config file in the same manner to what you've done here; https://github.com/DannyBen/bashly/discussions/405#discussioncomment-6630262 and then if the config file doesn't exist, it'll use the new config implementation to write to it. In the future I'm probably going to add to my script a validation that some config options exist and add only those if they don't.

I'm gonna continue to explore the usage and implementation to my current script and report back how it works!

**Reply by DannyBen** [reply]
> if you can set a default value using ${config[section.key]:-"default value"}

Yes. Not only for arrays - any variable.
See https://stackoverflow.com/a/2013589/413924

Also - get the new `config.sh` library from [the master branch](https://github.com/DannyBen/bashly/blob/master/lib/bashly/libraries/config/config.sh) - I think I fixed the issues.

**Reply by Zelaf** [reply]
What am I missing? Trying to use what `set_command.sh` is using.
```
write_config() {

## Default vaules
config[generate.file_name]=uptime-kuma-service-push
config[generate.folder_location]=push-scripts
## Make config file
touch $script_path/config.ini
## Save config
config_save $script_path/config.ini

}
```
Error:
```
uptime-kuma-service-push: line 222: generate.file_name: syntax error: invalid arithmetic operator (error token is ".file_name")
uptime-kuma-service-push: line 217: generate.folder_location: syntax error: invalid arithmetic operator (error token is ".folder_location")
```
Tried with double quotes and $ before config too, can't seem to get it.

**Reply by DannyBen** [reply]
You are missing many thing:

1. bash does not know what is `config`, since you did not load it from anywhere. Since you are creating your own config, you need to declare it as associative array.
2. I do not know what is `$script_path` and where it is coming from
3. There is no need to `touch` it first

```bash
write_config() {
  declare -A config

  ## Default vaules
  config[generate.file_name]=uptime-kuma-service-push
  config[generate.folder_location]=push-scripts
  
  ## Save config
  config_save "config.ini"
}
```

**Reply by Zelaf** [reply]
Yeah, should've given more context, sorry.
`$script_path` comes from the `initialize.sh` to be able to save the script to any location and still use the scripts folder.

After your explanation I noticed that I did work the scripts a bit wrong. I can't load the config the normal way because the config doesn't exist. I also can't write the config because I can't load the config because the file still doesn't exist.
```
## Variable for script location
script_path=$(dirname "$(readlink -f "$0")")

## Generate config file if it doesn't exists
[[ -f "config.ini" ]] || write_config

## Set config file
config_load $script_path/config.ini
```
Setting `declare -A config` solved that perfectly!

**Comment by DannyBen** [comment]
See #406 - with a new config library and updated example.

Since this is a breaking change, I need to consider the release plan more carefully, but you can already use this example by simply copying the [config.sh](https://github.com/DannyBen/bashly/pull/406/files#diff-13706a0c0bdacc2eae10c3e0c649fa354f2987e9391c876556345d240122031d) file from that PR.

Let me know how it works for you.

**Reply by Zelaf** [reply]
Just added this to my current project and it's working wonderfully!

Thank you so much for this, it'll help me a lot!

**Reply by DannyBen** [reply]
Thanks for the suggestion and for confirming it is working.
I will release a version with it soon.

**Comment by DannyBen** [comment]
@Zelaf - note:

1. I have decided to keep the old `config` library as is
2. I have added a new library called `ini` which is more or less what you are using now - with all `config_*` functions renamed to `ini_*`

You can either:

1. Continue using what you have - in this case, remove the `[@bashly-upgrade config]` magic comment from your config.sh file (or avoid using `bashly generate --upgrade`)
2. Take the new `ini` library, and update your code accordingly.

**Reply by Zelaf** [reply]
Wouldn't the naming scheme confuse new users? Maybe call it something along the lines of v2 or so? I mean, I don't know what the majority of Bashly users require, but I would feel like this new config.ini system is far superior.

Perhaps "advanced-config"?

**Reply by DannyBen** [reply]
No. I have tried to convert [`rush`](https://github.com/dannyben/rush-cli), which was made with bashly, to the new config and it was more work and the code was less to my liking.

Also think if an INI reading library was available in a "real" programming language, it would have been called "ini" and not "config".

There are some nuances in both libraries which give them both some strengths and weaknesses.

I am not releasing this yet, so still open for debate.

**Reply by DannyBen** [reply]
And - as a side note - it is perfectly legitimate to modify the library code once it is injected to your `src` directory. One of the reasons bashly places these files in the "userland", is exactly for this purpose.

Just remember to remove the magic comment.

**Reply by Zelaf** [reply]
Ah alright yeah, I see that viewpoint. I did do a commit just now for the documentation site, didn't catch on that you weren't going to release it yet.

What are you current thoughts around the library then?

There's no real spec for how INI files work, closest I could find was [this site](https://docs.fileformat.com/system/ini/) but even there it shows that there's no standards for comments, escape characters and duplicate names.

What would you personally like to change to be better suited for rush? 

In the end for me I just want to be able to read key values and differentiate between sections inside of the code so if I have several commands under different sections that could use the same key name it wouldn't be an issue.

> And - as a side note...

I'm aware! I just want it to be as "bashly" as possible. I really love this neat little software and it's opened up so many possibilities for me! I've always had a tough time with learning programming languages so using bashly to get a software like experience in script formats is amazing!

**Reply by Zelaf** [reply]
I do also want to mention that since my project is so young still I'm very flexible in how the ini configuration can be implemented. But I do feel the need to have section organisation so if my projects where to expand to need more config options it would be possible.

**Reply by DannyBen** [reply]
> What are you current thoughts around the library then?

Well. The real problems are:

1. Bash does not have first-class arrays. Using arrays in bash is always awkward. This is especially true if you want your scripts to pass `shellcheck`. I spent several hours chasing my tail making things either work, or pass shellcheck - not both... :)
2. Bash does not have a way to nest arrays.
3. The "classic" INI file structure requires nesting.

These problems mean that you can never have a fully intuitive and easy to use INI library in bash. At least this is what this experience taught me. Which, in turn means, that there *is* a place for multiple implementations - each with its own emphasis.

> There's no real spec for how INI files work

This does not concern me, partly for the above reasons. One cannot expect to have a full-blown INI handler in bash - and if someone happens to develop one - I will gladly replace bashly's INI library.

> What would you personally like to change to be better suited for rush?

The new INI library has these faults:

1. There are two sources of truth: The INI file and the associative array. This means that whenever an outside process updates the INI file, you are left with an outdated copy of the associative array. I observed this problem in real world scenario, when my rush CLI [called itself](https://github.com/DannyBen/rush-cli/blob/9da5e22ed171a2fe04360b982fa20c5f24ee65d2/src/get_command.sh#L15), with a function that updates the INI file.
2. Loading and saving files - in the old library, this is implicit. You just set `CONFIG_FILE` and move on. This makes it so the old library is working more like a "config object" - nobody needs to know the details (e.g. the file path) - just the API to its data. In the new library, you need to load and save with a path, every time you make any change. I solved this partially by making the save and load function take the path from `INI_FILE`, if it is set, but you still have to load and save explicitely.
3. The old library abstracted all the operations to bash functions. So it was easy to use. The new library just gives you the hash, and lets you use it as you see fit (e.g. default values, `unset` for deletions etc). Adding these methods to the new library creates two conflicting concepts. On a side note here, I updated the `config_get` function to allow a second value, to sere as default value.

**Reply by Zelaf** [reply]
> There are two sources of truth:...

This could definitely make things a lot more complicated and make the scripts have to arbitrarily update the configs after all outside processes.

> Loading and saving files...

I personally like this setup more, being able to configure on the fly if necessary in a way. But I suppose the way I'm thinking about it, it could be made with the original config library as well..

> The old library abstracted all the operations to bash functions....

This would definitely make things a lot easier in terms of scripting with the library in more advanced scripts.

Is there a possibility to add to the original config system a way to get the config options and separate the sections still? Doing `config_get section.key` with the old library and it seeing if the `section.` separator is there would allow me to easily rewrite my scripts and provide backwards compatability for others already using the library.
eg. `config_get section.key1` = `key1value` and `config_get key1` = `key1value`.

**Reply by DannyBen** [reply]
> Is there a possibility to add to the original config system a way to get the config options and separate the sections still

Well. I mentioned this problem, but probably too subtly. I will try to elaborate.

Having another function, named `config_get` is not a problem. It will just return the value from the hash. But then, the question is why have it at all? It does not make anything shorter: `config_get key` is similar in "weight" to `config[key]`.

The old `config_get` did more - it **loaded the file** if needed. We cannot implement it in the new way, since we do not know which file holds the config...

It will not be fully backwards compatible, since there is still the need to `config_load` manually.

Same goes for `config_set` - even more clearly.
The old `config_set` saved directly to the file. If there will be a new `config_set`, it will need to a) update the hash and b) save the file.... which again, we don't know which file.

So - the only reasonable (but unacceptable) way I found to do it, is to rely on the `CONFIG_FILE` variable, but then this makes the INI library awkward yet again - since some functions (load / save) accept either an explicit filename or `CONFIG_FILE` variable, while others *must* have `CONFIG_FILE`.

See the problem?

**Reply by Zelaf** [reply]
So I tried a bunch of things trying to load in the sections and using a regex separator to differentiate between a section and a key while also keeping backwards compatibility but couldn't get it to work at all. I'm far from anywhere near good at bash to get anything going but I got this so far and it *might* just work if it were to get some elbow grease.

```bash
config_get() {
  local section_key=$1
  local section_key_regex="^([^\.]+)\.([^\.]+)$"
  local section="${section_key%%.*}"
  local key="${section_key#*.}"
  local regex="^$key *= *(.+)$"
  local value=""

  config_init

  while IFS= read -r line || [ -n "$line" ]; do
    if [[ "$line" =~ $section_key_regex ]]; then
      current_section="${BASH_REMATCH[1]}"
      if [[ "$current_section" = "$section" ]]; then
        if [[ "$line" =~ $regex ]]; then
          value="${BASH_REMATCH[1]}"
          break
        fi
      fi
    fi
  done < "$CONFIG_FILE"

  echo "$value"
}
```

But that's only for the config_get function. I have no clue how to further do it for config_set function. I would love to be able to create a backwards compatible system where it completely uses the old config system but has support for detecting and separating between sections and if there's no `section.` put in place it'll just do the old `key = value`.

**Reply by DannyBen** [reply]
There is something nice and simple in loading everything to an array - this is the nice thing about the new approach.
I don't think we should go back to parsing the files directly with functions - this will probably get real ugly real fast.

It was also my initial attempt, but then we both realized that we should not invent "array access" when it already exists.

Another advantage of the array approach, is that it is faster. Load once, and use the values many times. Or update multiple values, and save once.

So still - given the bash constraints - I think both implementations have merit.

1. config library - quick, easy and intuitive key=value access
2. ini library - a more "formal" INI handling utility functions

**Reply by Zelaf** [reply]
I see, that does make a lot of sense, and faster processing is always a plus!

But I guess I still don't really understand the difference between arrays and, variables I guess the equivalent would be?

For the documentation however I think it would be best to give a bit clearer terms on the difference between the two. What would be practical examples for using the different libraries?

**Reply by DannyBen** [reply]
> I still don't really understand the difference between arrays and, variables I guess the equivalent would be?

Can you rephrase? I don't understand the question.

> For the documentation however I think it would be best to give a bit clearer terms on the difference

It is probably a matter of preference.
The bashly libraries are not the core of bashly. They are provided as complimentary extras.
While I agree that ideally there should be one INI handler, as we thoroughly discussed here - there is no such implementation that holds all the advantages, and the disadvantages of each could be substantial depending on the use case.

I do not like long documentation. Long documentation means it takes longer to learn. If it needs long documentation, I usually avoid the feature instead.

**Reply by DannyBen** [reply]
One more thing.

There *may* be a way for me to merge the two libraries, if we accept this feature reduction:
If we remove the explicit filename from the load/save functions - it might be easier to have both array-level access, and function access (e.g. `config_set key value` which will set it in the array and save the file).

Meaning - we use it like the old way, where the filename is dictated by `CONFIG_FILE`.

Thoughts?

**Reply by Zelaf** [reply]
> Can you rephrase? I don't understand the question.

I'm trying to say that I don't really understand what an array really is lol

> Meaning - we use it like the old way, where the filename is dictated by CONFIG_FILE.

I wouldn't mind this. Currently I load the filename in the `initialize.sh` in a way that would probably be more suited for the old config system.

`initialize.sh`
```
## Variable for script location
script_path=$(dirname "$(readlink -f "$0")")

## Generate config file if it doesn't exists
[[ -f "config.ini" ]] || write_config

## Set config file
config_load $script_path/config.ini
```
write_config.sh
```
write_config() {

    declare -A config

    ## Default vaules
    config[generate.file_name]=uptime-kuma-service-push
    config["generate.folder_location"]=push-scripts

    ## Save config
    config_save "$script_path/config.ini"

}
```
For my use case it probably wouldn't matter too much and since it's a new feature anyway why not try and focus on backwards compatibility? In the end the important feature for me are the sections, how they get implemented would be to what's best for Bashly as a whole.

**Reply by DannyBen** [reply]
> I'm trying to say that I don't really understand what an array really is lol

Oh - array is a list of variables (or literal values) that is treated as a whole.
Sample array in bash:

```bash
fruits=("Apple" "Banana" "Orange" "Grapes" "Mango")
echo "First fruit: ${fruits[0]}"
```

The new INI library uses **associative array** - which is also known as hash or dictionary in other languages.
In this case, the index of each item is a string (normally) and not a number:

```bash
config[options.username]
```

> `script_path=$(dirname "$(readlink -f "$0")")`

A shorter way: `SCRIPT_DIR="$(dirname "$0")"`
And I would also consider using uppercase for "super globals" like this to distinguish them from local vars, or from dynamically generated global vars (like the `ini` array).

> I wouldn't mind this

I will give it a shot.

**Reply by Zelaf** [reply]
> A shorter way: SCRIPT_DIR="$(dirname "$0")"

Will this work with symlinks too? My script is going to be installable to `/usr/local/bin/` so need to read where the symlink points too as well!

**Reply by DannyBen** [reply]
Put configs in `~/.config/`.
`dirname "$0"` will pring the dir of the symlink.

**Comment by DannyBen** [comment]
Ok - another attempt. Do you want to try it out?

### Highlights

1. No more direct access through the array. All access is done via dedicated functions.
2. File is auto-loaded and auto-saved, with the ability to force reload if someone needs to (e.g. the rush use case).
3. Does not load the file unless it is necessary (minimum possible IO operations).
4. Backwards compatible with the old config library

### `src/lib/config.sh`

<details>
  <summary>Show Code</summary>

```bash
## Config (INI) functions [@bashly-upgrade config]
## This file is a part of Bashly standard library
##
## Usage:
##
## - Set the global variable CONFIG_FILE to the path of your INI file somewhere
##   in your script (for example, in src/initialize.sh)
## - Use any of the following functions to access and manipulate the values.
## - INI sections are optional (i.e., sectionless key=value pairs are allowed).
##

## Show all the key-value pairs from your config file
config_show() {
  config_load

  for key in $(config_keys); do 
    echo "$key = ${config[$key]}"
  done
}

## Get a single value from the config file:
##
##   config_get login.email
##
## Get the value or a default one if it is not set:
##
##   config_get login.user guest
##
## Assign the result to a variable:
##
##   theme="$(config_get interface.theme)"
##
config_get() {
  local key="$1"
  local default_value="$2"
  
  config_load  
  echo "${config["$key"]:-$default_value}"
}

## Create/update a key=value pair:
##
##   config_set cloud.provider aws
##
config_set() {
  local key="$1"
  shift
  local value="$*"
  
  config_load
  config["$key"]="$value"
  config_save
}

## Delete a key=value pair:
##
##   config_del login.email
##
config_del() {
  local key="$1"
  
  config_load
  unset "config[$key]"
  config_save
}

## Get an array of all keys:
##
##   for k in $(config_keys); do
##     echo "- $k = $(config_get "$k")";
##   done
##
config_keys() {
  config_load

  local keys=("${!config[@]}")
  for a in "${keys[@]}"; do echo "$a"; done | sort
}

## Check if a key exists:
##
##   if config_has_key "key"; then
##     echo "key exists"
##   fi
##
config_has_key() {
  [[ $(config_get "$1") ]]
}

## Force-load from file
## This should normally not be called, unless you suspect that the INI file
## was modified by external means during the run of your script.
config_reload() {
  declare -g config_loaded=false
  config_load
}

## === PRIVATE - FOR INTERNAL USE ===

## Load an INI file (unless loaded) and populate the associative array
config_load() {
  [[ "$config_loaded" == "true" ]] && return

  declare -g CONFIG_FILE=${CONFIG_FILE:=config.ini}
  declare -g config_loaded=true
  declare -gA config

  local section=""
  local key=""
  local value=""
  local section_regex="^\[(.+)\]"
  local key_regex="^([^ =]+) *= *(.*) *$"
  local comment_regex="^;"

  [[ -f "$CONFIG_FILE" ]] || touch "$CONFIG_FILE"

  while IFS= read -r line; do
    if [[ $line =~ $comment_regex ]]; then
      continue
    elif [[ $line =~ $section_regex ]]; then
      section="${BASH_REMATCH[1]}."
    elif [[ $line =~ $key_regex ]]; then
      key="${BASH_REMATCH[1]}"
      value="${BASH_REMATCH[2]}"
      config["${section}${key}"]="$value"
    fi
  done <"$CONFIG_FILE"
}

## Save the associative array back to a file
config_save() {
  declare -gA config

  local current_section=""
  local has_free_keys=false

  rm -f "$CONFIG_FILE"

  for key in $(config_keys); do
    [[ $key == *.* ]] && continue
    has_free_keys=true
    value="${config[$key]}"
    echo "$key = $value" >>"$CONFIG_FILE"
  done

  [[ "${has_free_keys}" == "true" ]] && echo >>"$CONFIG_FILE"

  for key in $(config_keys); do
    [[ $key == *.* ]] || continue
    value="${config[$key]}"
    IFS="." read -r section_name key_name <<<"$key"

    if [[ "$current_section" != "$section_name" ]]; then
      [[ $current_section ]] && echo >>"$CONFIG_FILE"
      echo "[$section_name]" >>"$CONFIG_FILE"
      current_section="$section_name"
    fi
    
    echo "$key_name = $value" >>"$CONFIG_FILE"
  done
}
```

</details>

### Sample usage code

```bash
#!/usr/bin/env bash
source config.sh

config_set center Mr. Jones
config_set login.user Operations
config_set login.email paul@section.one
config_show

echo "$(config_get login.user)"

config_del login.user
config_show

echo "$(config_get login.user guest)"
```

**Reply by Zelaf** [reply]
I added it to my script and it calls and writes wonderfully.

My write_script function
```
write_config() {

    ## Default values
    config_set "generate.file_name" "uptime-kuma-service-push"
    config_set "generate.folder_location" "push-scripts"
    config_set "config.monitor_folder_location" "monitored-scripts"
    config_set "config.monitor_file_name" "uptime-kuma-service-push"
    config_set "config.systemd_until_installed" "false"

}
```

And it writes out everything in alphabetical order funnily enough!
```
[config]
monitor_file_name = uptime-kuma-service-push
monitor_folder_location = monitored-scripts
systemd_until_installed = false

[generate]
file_name = uptime-kuma-service-push
folder_location = push-scripts
```
Other than an error in my filter scripts because I forgot to change it everything worked flawlessly!

**Reply by Zelaf** [reply]
Also made a test for a non-sectioned key.

```
write_config() {

    ## Default values
    config_set "generate.file_name" "uptime-kuma-service-push"
    config_set "generate.folder_location" "push-scripts"
    config_set "config.monitor_folder_location" "monitored-scripts"
    config_set "config.monitor_file_name" "uptime-kuma-service-push"
    config_set "config.systemd_until_installed" "false"
    config_set "testkey" "testvalue"

}
```
```
testkey = testvalue

[config]
monitor_file_name = uptime-kuma-service-push
monitor_folder_location = monitored-scripts
systemd_until_installed = false

[generate]
file_name = uptime-kuma-service-push
folder_location = push-scripts
```
It worked great too! Do you think this can replace the original config system? What's the difference between this one and the one you've made previously that was incompatible?

**Reply by DannyBen** [reply]
> And it writes out everything in alphabetical order funnily enough!

Yes. This is a must, otherwise you might get the section twice as you saw before. This ensures that all keys in the same section are grouped together.

> Do you think this can replace the original config system?

Probably.

> What's the difference between this one and the one you've made previously that was incompatible?

The previous library had these features:

1. Loading and saving was manual and received explicit filename.
2. The developer had direct access to the array itself.  The array is also available here, but it is not "published" as part of the API - so we no longer have two sources of truth (well, we do, but the library is responsible for reconciling them).

I need to test it more, maybe make some more changes, and then we'll see.
Also I am considering layering the library:

1. Having a "raw" INI handler library (read/write INI).
2. Having the config library, which will use the INI library.

Not sure if it has advantages, or if it will confuse people.

**Reply by Zelaf** [reply]
Considering your previous naming just "INI" and remembering you named the other one "yaml" maybe it would be nice to have the two but name the config-library in style of "basic-config" or add it to the description that it's a bit more of a basic config system? Personally I don't see a reason to not directly go for this new INI one every time because it just seems better in my eyes.

**Reply by DannyBen** [reply]
Naming is not going to change.
`config` is compatible with the old config, and is an abstraction layer for INI functions, while the INI functions are in the `ini` library.

I think I will merge.

**Reply by Zelaf** [reply]
Alright, that's pretty cool then. I'm glad you found a solution that works for everyone! I'm gonna setup the new system from #409 to my config and see how it goes. Sadly I probably won't be using much of the array functions in practicality.

**Reply by DannyBen** [reply]
Are you going to use the `ini` library directly? or the `config`?

**Reply by Zelaf** [reply]
`config` mostly. But I did install both in-case I need the other library in the future. I am a bit curious about how I could implement a practical example for arrays in my script but I'm unsure if there's a need for it now that I actually understand what they are lol. It was load once but since my project is run with configs and do one thing there's not too much that it does in one go except for generating files.

**Reply by DannyBen** [reply]
Well - the `config` library *needs* the `ini` library - one uses the other now, as can be seen in this installation manifest:

https://github.com/DannyBen/bashly/blob/f8995c6c0fa4b41cac8b428ed70ebf01fcc6a6d2/lib/bashly/libraries/libraries.yml#L22-L28

**Reply by Zelaf** [reply]
I see well, good thing I installed both then! Will this be added to `bashly add` anytime soon?

**Reply by DannyBen** [reply]
> Will this be added to `bashly add` anytime soon?

Sure. I usually release on the weekend, but if it is important to you - I can release today/tomorrow.

**Comment by DannyBen** [comment]
Ok - I think this is it: https://github.com/DannyBen/bashly/pull/409

When you add the [config library](https://github.com/DannyBen/bashly/blob/0696cdd78bd93204f87a1b0f6df8ca5a9696f230/lib/bashly/libraries/config/config.sh), it will also add the [INI library](https://github.com/DannyBen/bashly/blob/0696cdd78bd93204f87a1b0f6df8ca5a9696f230/examples/ini/src/lib/ini.sh).

I think this is a good conclusion.

1. The config library is the library that most people who need INI file access should use.
2. If they want to use the array itself for whatever reason, they can just use the low level INI library instead.

Makes sense?

**Reply by Zelaf** [reply]
I've implemented it to my current script and it seems to be working wonders! Thanks a bunch! :heart:

**Reply by DannyBen** [reply]
Thanks for your involvement. I will release on the weekend.
